File: apps/android/app/src/test/java/app/murmur/android/SettingsParityTest.kt
package app.murmur.android

import android.content.Context
import app.murmur.android.settings.FormattingMode
import app.murmur.android.settings.MurmurSettings
import app.murmur.android.settings.SettingsRanges
import app.murmur.android.settings.SettingsStore
import app.murmur.android.settings.SttPresets
import app.murmur.android.settings.Tone
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import java.io.File

/**
 * The Android side of the cross-platform settings contract
 * (apps/desktop/tests/fixtures/settings-parity.json): the names, defaults and bounds both apps
 * ship for everything a person can set on either, and the speech provider presets. The desktop
 * schema is pinned to the same file by its settings-parity test, so a default changed on one app
 * without the other fails one of the two.
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [35])
class SettingsParityTest {
    private val contract: JSONObject by lazy {
        val file = listOf(
            File("../../../apps/desktop/tests/fixtures/settings-parity.json"),
            File("../../apps/desktop/tests/fixtures/settings-parity.json"),
            File("apps/desktop/tests/fixtures/settings-parity.json")
        ).firstOrNull { it.exists() }
            ?: error("settings-parity.json not found from ${File(".").absolutePath}")
        JSONObject(file.readText())
    }

    private fun JSONArray.strings(): List<String> = (0 until length()).map { getString(it) }
    private fun JSONArray.ints(): List<Int> = (0 until length()).map { getInt(it) }

    @Test
    fun `the defaults are the ones both apps ship`() {
        val d = contract.getJSONObject("defaults")
        val s = MurmurSettings()
        assertEquals(d.getInt("sttTimeoutMs"), s.sttTimeoutMs)
        assertEquals(d.getInt("llmTimeoutMs"), s.llmTimeoutMs)
        assertEquals(d.getBoolean("limitDuration"), s.limitDuration)
        assertEquals(d.getInt("maxDurationSec"), s.maxDurationSec)
        assertEquals(d.getBoolean("keepRecordings"), s.keepRecordings)
        assertEquals(d.getBoolean("useDictionaryPrompt"), s.useDictionaryPrompt)
        assertEquals(d.getBoolean("showLatencyInHistory"), s.showLatencyInHistory)
        assertEquals(d.getBoolean("buttonShadow"), s.buttonShadow)
        assertEquals(d.getString("formattingMode"), s.formattingMode.id)
        assertEquals(d.getString("tone"), s.tone.id)
        assertEquals(d.getBoolean("trailingSpace"), s.trailingSpace)
        assertEquals(d.getString("instructions"), s.llmInstructions)
        assertEquals(d.getString("language"), s.language)
        assertEquals(d.getString("sttKind"), s.sttKind.id)
        assertEquals(d.getString("sttPresetId"), s.sttPresetId)
        assertEquals(d.getBoolean("llmSameAsStt"), s.llmSameAsStt)
        assertEquals(d.getBoolean("updateAutoCheck"), s.updateAutoCheck)
        assertEquals(d.getBoolean("updateAutoInstall"), s.updateAutoInstall)
        assertEquals(d.getBoolean("updateIncludePrereleases"), s.updateIncludePrereleases)
        assertEquals(d.getBoolean("sounds"), s.sounds)
        assertEquals(d.getDouble("soundVolume"), s.soundVolume.toDouble(), 0.0001)
        val k = s.keyboard
        assertEquals(d.getString("overlayPosition"), k.overlayPosition.id)
        assertEquals(d.getBoolean("showOverlayWhenIdle"), k.showOverlayWhenIdle)
        assertEquals(d.getJSONArray("pushToTalk").ints(), k.pushToTalk)
        assertEquals(d.getJSONArray("handsFree").ints(), k.handsFree)
        assertEquals(d.getString("handsFreeTrigger"), k.handsFreeTrigger.id)
        assertEquals(d.getInt("tapThresholdMs"), k.tapThresholdMs)
        assertEquals(d.getInt("doubleTapWindowMs"), k.doubleTapWindowMs)
        assertEquals(d.getBoolean("sideSensitive"), k.sideSensitive)
        assertEquals(d.getBoolean("escapeCancels"), k.escapeCancels)
        // The session cap is off, so a session runs until the user stops it.
        assertNull(s.sessionDurationLimitSec)
        assertEquals(300, s.copy(limitDuration = true).sessionDurationLimitSec)
    }

    @Test
    fun `the bounds are the ones both apps clamp to`() {
        val r = contract.getJSONObject("ranges")
        fun range(name: String): IntRange = r.getJSONObject(name).let { it.getInt("min")..it.getInt("max") }
        assertEquals(range("sttTimeoutMs"), SettingsRanges.STT_TIMEOUT_MS)
        assertEquals(range("llmTimeoutMs"), SettingsRanges.LLM_TIMEOUT_MS)
        assertEquals(range("maxDurationSec"), SettingsRanges.MAX_DURATION_SEC)
    }

    @Test
    fun `the speech provider presets name the same connections`() {
        val expected = contract.getJSONArray("sttPresets")
        assertEquals(expected.length(), SttPresets.ALL.size)
        for (i in 0 until expected.length()) {
            val e = expected.getJSONObject(i)
            val p = SttPresets.ALL[i]
            assertEquals(e.getString("id"), p.id)
            assertEquals(e.getString("kind"), p.kind.id)
            assertEquals(e.getString("baseUrl"), p.baseUrl)
            assertEquals(e.getString("defaultModel"), p.defaultModel)
            assertEquals(e.getJSONArray("models").strings(), p.models)
            assertEquals(e.getBoolean("requiresKey"), p.requiresKey)
            assertEquals(e.getBoolean("local"), p.local)
        }
        assertEquals("an unknown id is Custom", SttPresets.CUSTOM, SttPresets.find("nope").id)
    }

    @Test
    fun `the style vocabularies match`() {
        val rule = contract.getJSONObject("appRule")
        assertEquals(rule.getJSONArray("tones").strings(), Tone.entries.map { it.id })
        assertEquals(rule.getJSONArray("modes").strings(), FormattingMode.entries.map { it.id })
    }

    @Test
    fun `the store clamps a stored value into the bounds instead of running with it`() {
        val context = RuntimeEnvironment.getApplication()
        val prefs = context.getSharedPreferences("murmur_settings", Context.MODE_PRIVATE)
        prefs.edit().clear()
            .putInt("sttTimeoutMs", 999_000)
            .putInt("llmTimeoutMs", 250)
            .putInt("maxDurationSec", 0)
            .putInt("parityMigration", 1)
            .commit()
        val store = SettingsStore(context)
        assertEquals(120_000, store.get().sttTimeoutMs)
        assertEquals(1_000, store.get().llmTimeoutMs)
        assertEquals(5, store.get().maxDurationSec)
        // Writes are clamped too, whatever a screen hands over.
        store.update { it.copy(sttTimeoutMs = 1, llmTimeoutMs = 90_000, maxDurationSec = 5_000) }
        assertEquals(2_000, prefs.getInt("sttTimeoutMs", 0))
        assertEquals(60_000, prefs.getInt("llmTimeoutMs", 0))
        assertEquals(1_800, prefs.getInt("maxDurationSec", 0))
    }
}

File: apps/desktop/tests/settings-parity.test.ts
import { readFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'
import {
  LLM_INSTRUCTIONS_MAX,
  SETTINGS_RANGES,
  clampToRange,
  defaultSettings,
  formattingModeSchema,
  toneSchema
} from '@shared/settings'
import { STT_PRESETS } from '@core/stt/presets'

/**
 * The desktop side of the cross-platform settings contract (tests/fixtures/settings-parity.json).
 * The Android store pins itself to the same file (SettingsParityTest), so a default or a bound
 * changed here without the phone following fails one of the two.
 */
const contract = JSON.parse(
  readFileSync(new URL('./fixtures/settings-parity.json', import.meta.url), 'utf8')
) as {
  defaults: Record<string, unknown>
  ranges: Record<string, { min: number; max: number }>
  sttPresets: Array<Record<string, unknown>>
  snippetPlaceholders: string[]
  appRule: { tones: string[]; modes: string[] }
}

describe('settings parity contract', () => {
  it('the defaults are the ones both apps ship', () => {
    const s = defaultSettings()
    expect({
      sttTimeoutMs: s.stt.timeoutMs,
      llmTimeoutMs: s.formatting.llm.timeoutMs,
      limitDuration: s.audio.limitDuration,
      maxDurationSec: s.audio.maxDurationSec,
      keepRecordings: s.audio.keepRecordings,
      useDictionaryPrompt: s.stt.useDictionaryPrompt,
      showLatencyInHistory: s.general.showLatencyInHistory,
      buttonShadow: s.general.buttonShadow,
      formattingMode: s.formatting.mode,
      tone: s.formatting.tone,
      trailingSpace: s.formatting.trailingSpace,
      instructions: s.formatting.instructions,
      language: s.stt.language,
      sttKind: s.stt.kind,
      sttPresetId: s.stt.presetId,
      llmSameAsStt: s.formatting.llm.sameAsStt,
      updateAutoCheck: s.updates.autoCheck,
      updateAutoInstall: s.updates.autoInstall,
      updateIncludePrereleases: s.updates.includePrereleases
    }).toEqual(contract.defaults)
  })

  it('the bounds are the ones both apps clamp to', () => {
    expect({
      ...SETTINGS_RANGES,
      instructionsLength: { min: 0, max: LLM_INSTRUCTIONS_MAX }
    }).toEqual(contract.ranges)
  })

  it('the speech provider presets name the same connections', () => {
    expect(
      STT_PRESETS.map((p) => ({
        id: p.id,
        kind: p.kind,
        baseUrl: p.baseUrl,
        defaultModel: p.defaultModel,
        models: p.models,
        requiresKey: p.requiresKey,
        local: !!p.local
      }))
    ).toEqual(contract.sttPresets)
  })

  it('the style vocabularies match', () => {
    expect(toneSchema.options).toEqual(contract.appRule.tones)
    expect(formattingModeSchema.options).toEqual(contract.appRule.modes)
  })

  it('an input clamps into the range instead of handing the schema a value it would reject', () => {
    // A rejected value drops its whole settings section to defaults (parseSettings salvages per
    // section), which for the transcription timeout means losing the provider URL and key.
    expect(clampToRange(121_000, SETTINGS_RANGES.sttTimeoutMs)).toBe(120_000)
    expect(clampToRange(1_000, SETTINGS_RANGES.sttTimeoutMs)).toBe(2_000)
    expect(clampToRange(45_000, SETTINGS_RANGES.sttTimeoutMs)).toBe(45_000)
    expect(clampToRange(Number.NaN, SETTINGS_RANGES.llmTimeoutMs)).toBe(1_000)
    expect(clampToRange(0, SETTINGS_RANGES.maxDurationSec)).toBe(5)
    expect(clampToRange(1_801, SETTINGS_RANGES.maxDurationSec)).toBe(1_800)
  })
})
